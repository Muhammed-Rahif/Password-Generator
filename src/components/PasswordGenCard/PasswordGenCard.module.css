.card {
  display: flex;
  justify-content: space-evenly;
  align-items: center;
  flex-direction: column;
}

.cardContent {
  display: flex;
  justify-content: space-between;
  margin-bottom: 0rem;
}

.cardHeader {
  display: flex;
  justify-content: center;
  align-items: center;
}

.cardTitle {
  text-align: center;
}

strong {
  text-transform: capitalize;
}

.copyBtn {
  margin-right: 0.5rem;
}

.generateBtn {
}

@media only screen and (max-width: 500px) {
  .card {
    min-height: 45vh;
    width: 93vw !important;
  }
  .password {
    width: 60vw !important;
  }
}
