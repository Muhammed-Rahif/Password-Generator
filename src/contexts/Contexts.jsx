import React, { createContext, useState } from "react";

const isDeviceLightTheme =
  window.matchMedia && window.matchMedia("(prefers-color-scheme: light)").matches;

export const LightThemeContext = createContext(isDeviceLightTheme);

function Contexts({ children }) {
  const [lightTheme, setLightTheme] = useState(isDeviceLightTheme);

  return (
    <LightThemeContext.Provider value={{ lightTheme, setLightTheme }}>
      {children}
    </LightThemeContext.Provider>
  );
}

export default Contexts;
