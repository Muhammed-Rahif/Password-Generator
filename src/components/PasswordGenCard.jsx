import React, { useContext, useState } from "react";
import styles from "./PasswordGenCard.module.css";
import { TextField, Card, Button, CardHeader, CardContent, H5 } from "ui-neumorphism";
import passwordDetails from "zxcvbn";
import { LightThemeContext } from "../contexts/Contexts";
import passwordGenerator from "generate-password";

function PasswordGenCard(props) {
  const [password, setPassword] = useState("");
  const [passwordStrong, setPasswordStrong] = useState("Enter a password or generate!");
  const [passwordLength, setPasswordLength] = useState(8);
  const [copyBtnText, setCopyBtnText] = useState("Copy");

  const { lightTheme } = useContext(LightThemeContext);

  const getPassStatusByScore = password => {
    switch (password) {
      case 0:
        return "Too weak!";
      case 1:
        return "Weak!";
      case 2:
        return "Medium!";
      case 3:
        return "Good!";
      case 4:
        return "Strong!";
      default:
        return "Something wrong!";
    }
  };

  const handlePasswordChange = password => {
    document.getElementById("password-input").value = password;
    setPassword(password);
    setPasswordStrong(getPassStatusByScore(passwordDetails(password).score));
  };

  return (
    <Card dark={!lightTheme} className={styles.card}>
      <CardHeader className={styles.cardHeader}>
        <H5 className={styles.cardTitle}>
          <small>Generate a super</small>
          <br />
          <strong>strong password!</strong>
        </H5>
      </CardHeader>
      <CardContent className={styles.cardContent}>
        <TextField
          placeholder="Click generate..."
          hint={passwordStrong}
          counter={passwordLength}
          onChange={e => handlePasswordChange(e.value)}
          value={password}
          id="password-input"
          uncontrolled
          className={styles.password}
        />
        <TextField
          onChange={e => {
            let length = parseInt(e.value);
            if (!length) return setPasswordLength(0);
            setPasswordLength(length > 64 ? 64 : length);
          }}
          value={passwordLength.toString()}
          className={styles.passLengthInput}
          width={50}
          uncontrolled
        />
      </CardContent>
      <CardContent>
        <Button
          className={styles.copyBtn}
          title="Password length"
          onClick={() => {
            setCopyBtnText("Copied!");
            navigator.clipboard.writeText(password);
            setTimeout(() => {
              setCopyBtnText("Copy");
            }, 2500);
          }}
        >
          {copyBtnText}
        </Button>
        <Button
          className={styles.generateBtn}
          title="Password length"
          onClick={() => {
            setPasswordLength(passwordLength < 12 ? 12 : passwordLength);
            let strongPass = passwordGenerator.generate({
              length: passwordLength < 12 ? 12 : passwordLength,
              numbers: true,
              symbols: true,
            });
            handlePasswordChange(strongPass);
          }}
        >
          Generate
        </Button>
      </CardContent>
    </Card>
  );
}

export default PasswordGenCard;
