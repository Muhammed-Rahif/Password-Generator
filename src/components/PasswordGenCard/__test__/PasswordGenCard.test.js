import React from "react";
import renderer from "react-test-renderer";
import Contexts from "../../../contexts/Contexts";
import PasswordGenCard from "../PasswordGenCard";

describe("password generating card", () => {
  const passwordGenCardComponent = renderer.create(
    <Contexts>
      <PasswordGenCard />
    </Contexts>
  );
  let componentTree = passwordGenCardComponent.toJSON();

  it("should render without crashing", () => {
    expect(componentTree).toMatchSnapshot();
  });
});
