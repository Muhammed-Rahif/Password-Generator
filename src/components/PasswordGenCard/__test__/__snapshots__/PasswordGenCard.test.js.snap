// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`password generating card should render without crashing 1`] = `
<div
  className="_28vIk _2bGnb _2nMGC card"
  style={Object {}}
>
  <div
    className="_2TuLn _1Vkyx cardHeader"
    style={Object {}}
  >
    <h5
      className="_3yPwv _2TwOi _11JjB _3PUqy cardTitle"
      style={Object {}}
    >
      <small>
        Generate a super
      </small>
      <br />
      <strong>
        strong password!
      </strong>
    </h5>
  </div>
  <div
    className="_2So3v _BClU8 cardContent"
    style={Object {}}
  >
    <div
      className="_2xbD- password"
      style={Object {}}
    >
      <div
        className="_fWLwS"
      >
        <input
          className="_U6nBC _1a45g "
          disabled={false}
          id="password-input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Click generate..."
          style={
            Object {
              "height": "undefinedpx",
              "minHeight": "undefinedpx",
              "width": "undefinedpx",
            }
          }
          type="text"
          value=""
        />
        <div
          className="_3t8tT"
        >
          <div
            className="_3yPwv _2HbLl _1QRSN _3PUqy _2Lw0g"
            style={Object {}}
          >
            Enter a password or generate!
          </div>
          <div
            className="_3yPwv _2HbLl _1QRSN _3PUqy _2MZzb"
            style={Object {}}
          >
            0
            /
            8
          </div>
        </div>
      </div>
    </div>
    <div
      className="_2xbD- passLengthInput"
      style={Object {}}
    >
      <div
        className="_fWLwS"
      >
        <input
          className="_U6nBC _1a45g "
          disabled={false}
          id="_50310_"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          style={
            Object {
              "height": "undefinedpx",
              "minHeight": "undefinedpx",
              "width": "50px",
            }
          }
          type="text"
          value="8"
        />
        <div
          className="_3t8tT"
        >
          <div
            className="_3yPwv _2HbLl _1QRSN _3PUqy _2Lw0g"
            style={Object {}}
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="_2So3v _BClU8 "
    style={Object {}}
  >
    <div
      className="_cL4RU _17_OS  _3W7Om _3tLvr copyBtn"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "var(--success)",
          "border": null,
          "color": undefined,
        }
      }
    >
      <button
        className="_3lLt8"
      >
        Copy
      </button>
    </div>
    <div
      className="_cL4RU _17_OS  _3W7Om _3tLvr generateBtn"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "var(--info)",
          "border": null,
          "color": undefined,
        }
      }
    >
      <button
        className="_3lLt8"
      >
        Generate
      </button>
    </div>
  </div>
</div>
`;
